shared/Microsoft.Extensions.Certificates.Generation.Sources/CertificateManager.cs

```diff
@@ -0,0 +1,435 @@
+                    Encoding.ASCII.GetBytes(AspNetIdentityOidFriendlyName)),
+                critical: false);
+
+            extensions.Add(basicConstraints);
+            extensions.Add(keyUsage);
+            extensions.Add(enhancedKeyUsage);
+            extensions.Add(aspNetIdentityExtension);
+
+            var certificate = CreateSelfSignedCertificate(subject, extensions, notBefore, notAfter);
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                certificate.FriendlyName = AspNetIdentityOidFriendlyName;
+            }
+
+            return certificate;
+        }
+
+        public X509Certificate2 CreateSelfSignedCertificate(
+            X500DistinguishedName subject,
+            IEnumerable<X509Extension> extensions,
+            DateTimeOffset notBefore,
+            DateTimeOffset notAfter)
+        {
+            var key = CreateKeyMaterial(RSAMinimumKeySizeInBits);
+
+            var request = new CertificateRequest(subject, key, HashAlgorithmName.SHA256, RSASignaturePadding.Pkcs1);
+            foreach (var extension in extensions)
+            {
+                request.CertificateExtensions.Add(extension);
+            }
+
+            return request.CreateSelfSigned(notBefore, notAfter);
+
+            RSA CreateKeyMaterial(int minimumKeySize)
+            {
+                var rsa = RSA.Create(minimumKeySize);
+                if (rsa.KeySize < minimumKeySize)
+                {
+                    throw new InvalidOperationException($"Failed to create a key with a size of {minimumKeySize} bits");
+                }
+
+                return rsa;
+            }
+        }
+
+        public X509Certificate2 SaveCertificateInStore(X509Certificate2 certificate, StoreName name, StoreLocation location)
+        {
+            var imported = certificate;
+            if (!RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
+            {
+                // On non OSX systems we need to export the certificate and import it so that the transient
+                // key that we generated gets persisted.
+                var export = certificate.Export(X509ContentType.Pkcs12, "");
+                imported = new X509Certificate2(export, "", X509KeyStorageFlags.PersistKeySet | X509KeyStorageFlags.Exportable);
+                Array.Clear(export, 0, export.Length);
+            }
+
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                imported.FriendlyName = certificate.FriendlyName;
+            }
+
+            using (var store = new X509Store(name, location))
+            {
+                store.Open(OpenFlags.ReadWrite);
+                store.Add(imported);
+                store.Close();
+            };
+
+            return imported;
+        }
+
+        public void ExportCertificate(X509Certificate2 certificate, string path, bool includePrivateKey, string password)
+        {
+            if (Path.GetDirectoryName(path) != "")
```

Comment:
Should this be an error if `string.IsNullOrEmpty(Path.GetDirectoryName(path))` or `!Path.IsRooted(path)`? Seems like it would be better to require absolute paths, otherwise you get output relative to Directory.GetCurrentDirectory(), which changes based on how you invoke your process (IIS, IIS Express, command-line, VS, etc).

Reply:
I'm going to put all those semantics on the actual exe we will build. I want to keep the library flexible